crates/build/src/docker.rs

```diff
@@ -0,0 +1,563 @@
+    );
+    build_result.target_directory = new_path;
+
+    let new_path = build_result.dest_wasm.as_ref().map(|p| {
+        host_folder.join(
+            p.as_path()
+                .strip_prefix(MOUNT_DIR)
+                .expect("cannot strip prefix"),
+        )
+    });
+    build_result.dest_wasm = new_path;
+
+    build_result.metadata_result.as_mut().map(|mut m| {
+        m.dest_bundle = host_folder.join(
+            m.dest_bundle
+                .as_path()
+                .strip_prefix(MOUNT_DIR)
+                .expect("cannot strip prefix"),
+        );
+        m.dest_metadata = host_folder.join(
+            m.dest_metadata
+                .as_path()
+                .strip_prefix(MOUNT_DIR)
+                .expect("cannot strip prefix"),
+        );
+        m
+    });
+    Ok(build_result)
+}
+
+/// Overwrites `build_result` and `image` fields in the metadata
+fn update_metadata(
+    build_result: &BuildResult,
+    verbosity: &Verbosity,
+    build_steps: &mut BuildSteps,
+    build_image: &ImageSummary,
+) -> Result<()> {
+    if let Some(metadata_artifacts) = &build_result.metadata_result {
+        let mut metadata = ContractMetadata::load(&metadata_artifacts.dest_bundle)?;
+        // find alternative unique identifier of the image, otherwise grab the digest
+        let image_tag = match build_image
+            .repo_tags
+            .iter()
+            .find(|t| !t.ends_with("latest"))
+        {
+            Some(tag) => tag.to_owned(),
+            None => build_image.id.clone(),
+        };
+
+        metadata.image = Some(image_tag);
+
+        crate::metadata::write_metadata(
+            metadata_artifacts,
+            metadata,
+            build_steps,
+            verbosity,
+        )?;
+    }
+    Ok(())
+}
+
+/// Creates the container and executed the build inside it
+async fn run_build(
+    build_args: String,
+    build_image: &ImageSummary,
+    build_result_path: &PathBuf,
+    contract_name: &str,
+    host_folder: &Path,
+    verbosity: &Verbosity,
+    build_steps: &mut BuildSteps,
+) -> Result<()> {
+    let client = Docker::connect_with_socket_defaults()?;
+
+    let mut entrypoint = vec!["/bin/bash".to_string(), "-c".to_string()];
+    let b_path_str = build_result_path
+        .as_os_str()
+        .to_str()
+        .context("Cannot convert Os String to String")?;
+    let mut cmds = vec![format!(
+        "mkdir -p target && cargo contract build {} --output-json > {}",
+        build_args, b_path_str
+    )];
+
+    entrypoint.append(&mut cmds);
+
+    // in order to optimise the container usage
+    // we are hashing the inputted command
+    // in order to reuse the container for the same permutation of arguments
+    let mut s = DefaultHasher::new();
+    entrypoint.hash(&mut s);
+    let digest = s.finish();
+    // taking the first 5 digits to be a unique identifier
+    let digest_code: String = digest.to_string().chars().take(5).collect();
```

Comment:
This was causing some issues for me because if there already exists a container with the id derived from the entrypoint from a *different* image, it will just use that so the image is effectively ignored and it will just use that container.

Reply:
Good spot. I will include the image name in the calculation of the digest.

Reply:
Dumb question perhaps: why do the hashing at all and not just use the concatenated image + contract name, does it have to be a unique for different argument permutations? What would happen if they just use the same container?

Reply:
You can not use the same container, because upon creation, the entry point command becomes part of the container. So if you need to run a build with different set of args, you need to create a new container. So the idea is that we spawn a new container for a new permutation of args, and then reuse it. The reason for that was discussed before, because we cannot use `docker exec` since there is a limitation in RAM when running on Apple Silicon because of emulation overhead.

Reply:
> there is a limitation in RAM when running on Apple Silicon because of emulation overhead.

I haven't experience that problem. Can you post steps to reproduce @SkymanOne ?

Reply:
> > there is a limitation in RAM when running on Apple Silicon because of emulation overhead.
> 
> I haven't experience that problem. Can you post steps to reproduce @SkymanOne ?

See previous discussion: https://github.com/paritytech/cargo-contract/pull/1148#discussion_r1223379844